### DISCUSSION TITLE
The state of array gets reset to its initial state randomly upon porting to App Router

### DISCUSSION THREAD

**Discussion by valehh** [vercel/next.js#57076] [Help]
### Summary

Hi, I am upgrading my app from pages to App Router and I have encountered a problem I cannot figure out.

**The code:** I have a google map that I have loaded dynamically. There are some some polygons on the map. The user can select multiple polygons on the map by clicking/tapping them. This is simple code that was working well in pages.

const [selectedCom, setSelectedCom] = useState( [] );

function handleSelectedPolygonsChange(id) { //this is invoked onClick of a polygon
    setSelectedCom( [...selectedCom, id] );
}

**The problem:** That feature no longer works as expected. The selectedCom randomly becomes empty and I lose all selected ids. 

- There is no race condition as I am not using that variable anywhere else.
- I tried initializing the state const with "aa" and "bb" instead of empty array ( useState( ["aa", "bb"] ) to see what happens. I noticed that the selectedCom array randomly becomes ["aa","bb"] and I lose all the selected ids.
- It is very random. If I count until 5 and keep clicking once in a while I lose the state and setSelectedCom resets to its initial state.

I am thinking this might be a caching issue where randomly a cache is used? Is there a cache for state in app router? (there is no data fetching involved here). What else could be a cause? The Pages approach was working like a charm. I am not using any framework feature here so it is a very weird issue.

Thanks.

**Comment by icyJoseph** [comment]
This sounds like stale closure on `handleSelectedPolygonsChange`, but without seeing any code that's my only guess.

Try with:

```ts
function handleSelectedPolygonsChange(id) { //this is invoked onClick of a polygon
setSelectedCom( prev => [...prev, id] );
}
```
